pkg/server/infrastructure/datastore/mysql/mysql.go

```diff
@@ -47,40 +64,249 @@ func New(ctx context.Context, cfg datastore.Config) (datastore.DataStore, error)
+			}
+			for _, deleteEntity := range entities[:i] {
+				if _, exit := notRollback[deleteEntity.PrimaryKey()]; !exit {
+					if err := m.Delete(ctx, deleteEntity); err != nil {
+						if !errors.Is(err, datastore.ErrRecordNotExist) {
+							klog.Errorf("rollback delete entity failure %w", err)
+						}
+					}
+				}
+			}
+			return datastore.NewDBError(fmt.Errorf("save entities occur error, %w", err))
+		}
+	}
+	return nil
 }
 
 // Get get data model
 func (m *mysql) Get(ctx context.Context, entity datastore.Entity) error {
-	return fmt.Errorf("method to be implemented")
+	if entity.PrimaryKey() == "" {
+		return datastore.ErrPrimaryEmpty
+	}
+	if entity.TableName() == "" {
+		return datastore.ErrTableNameEmpty
+	}
+
+	if dbGet := m.client.First(entity); dbGet.Error != nil {
+		if errors.Is(dbGet.Error, gorm.ErrRecordNotFound) {
+			return datastore.ErrRecordNotExist
+		}
+		return datastore.NewDBError(dbGet.Error)
+	}
+	return nil
 }
 
 // Put update data model
 func (m *mysql) Put(ctx context.Context, entity datastore.Entity) error {
-	return fmt.Errorf("method to be implemented")
+	if entity.PrimaryKey() == "" {
+		return datastore.ErrPrimaryEmpty
+	}
+	if entity.TableName() == "" {
+		return datastore.ErrTableNameEmpty
+	}
+	entity.SetUpdateTime(time.Now())
+	if dbPut := m.client.Model(entity).Updates(entity); dbPut.Error != nil {
+		if errors.Is(dbPut.Error, gorm.ErrRecordNotFound) {
+			return datastore.ErrRecordNotExist
+		}
+		return datastore.NewDBError(dbPut.Error)
+	}
+	return nil
 }
 
 // IsExist determine whether data exists.
 func (m *mysql) IsExist(ctx context.Context, entity datastore.Entity) (bool, error) {
-	return true, fmt.Errorf("method to be implemented")
+	if entity.PrimaryKey() == "" {
+		return false, datastore.ErrPrimaryEmpty
+	}
+	if entity.TableName() == "" {
+		return false, datastore.ErrTableNameEmpty
+	}
+
+	if dbExist := m.client.First(entity); dbExist.Error != nil {
+		if errors.Is(dbExist.Error, gorm.ErrRecordNotFound) {
+			return false, nil
+		}
+		return false, datastore.NewDBError(dbExist.Error)
+	}
+
+	return true, nil
 }
 
 // Delete delete data
 func (m *mysql) Delete(ctx context.Context, entity datastore.Entity) error {
-	return fmt.Errorf("method to be implemented")
+	if entity.PrimaryKey() == "" {
+		return datastore.ErrPrimaryEmpty
+	}
+	if entity.TableName() == "" {
+		return datastore.ErrTableNameEmpty
+	}
+	// check entity is exist
+	if err := m.Get(ctx, entity); err != nil {
+		return err
+	}
+
+	if dbDelete := m.client.Unscoped().Model(entity).Delete(entity); dbDelete.Error != nil {
+		klog.Errorf("delete document failure %w", dbDelete.Error)
+		return datastore.NewDBError(dbDelete.Error)
+	}
+
+	return nil
+}
+
+func _applyFilterOptions(clauses []clause.Expression, filterOptions datastore.FilterOptions) []clause.Expression {
+	for _, queryOp := range filterOptions.Queries {
+		clauses = append(clauses, clause.Like{
+			Column: strings.ToLower(queryOp.Key),
```

Comment:
Why we use `strings.ToLower` is implicit. I know you are casting all column name to lower cases but if someone doesn't know the design, this functions will be confusing. Let's alias it and use alias. 
```go
var toColumnName = strings.ToLower
```

Reply:
Yeah, its better to alias it and then using it, Also IMO adding some comments is better.

Reply:
Done !!